### PR TITLE
Error while putting document #6847

### LINKED ISSUE #6847: Error while  putting document
I've been getting a few of these errors recently from the live server. 
Here's an example:
```
Error Message: teammates.common.util.Logger severe: Operation did not succeed in time: putting document Document(documentId=abc@purdue.edu%CGT27208%2074176116, fields=[Field(name=courseId, value=CGT27208, type=TEXT), Field(name=searchableText, value=CGT27208,HCDD Learning Studio 3,abc@purdue.edu,ABC,Team 1,None, type=TEXT), Field(name=studentAttribute, value={
"googleId": "",
"email": "abc@purdue.edu",
"course": "CGT27208",
"name": "ABC",
"lastName": "DEF",
"comments": "",
"team": "Team 1",
"section": "None",
"key": "abc@purdue.edu%CGT27208%2074176116"
}, type=TEXT)], rank=195923046) into search index student.
```
v5.99

### PR DESCRIPTION
**Pull request by whipermr5**
Fixes #6847

**Outline of Solution**

- Wrap persistence check in if block
- Use exponential backoff to handle transient backend errors (follow [Google](https://cloud.google.com/appengine/docs/standard/java/search/#Java_Putting_documents_in_an_index))

### CODE REVIEW

File: src/main/java/teammates/storage/search/SearchManager.java [hunk 1]
```diff
@@ -41,34 +42,42 @@ private SearchManager() {
...
-            if (!StatusCode.TRANSIENT_ERROR.equals(e.getOperationResult().getCode())) {
-                log.severe(String.format(ERROR_NON_TRANSIENT_BACKEND_ISSUE, document, indexName)
-                           + TeammatesException.toStringWithStackTrace(e));
+
+        int delay = 2;
+        for (int attempts = 0; attempts < MAX_RETRIES; attempts++) {
+            try {
+                PutResponse result = index.put(document);
+
+                if (Config.PERSISTENCE_CHECK_DURATION > 0) {
```
Comment (damithc) [inline src/main/java/teammates/storage/search/SearchManager.java]: Reverse the condition (i.e. make it a guard clause) to reduce nesting.

File: src/main/java/teammates/storage/search/SearchManager.java [hunk 2]
```diff
@@ -49,23 +49,26 @@ public static void putDocument(String indexName, Document document) {
...
-                        // check before incrementing to avoid boundary case problem
-                        if (!isSuccessful) {
-                            elapsedTime += ThreadHelper.WAIT_DURATION;
-                        }
-                    }
-                    if (elapsedTime >= Config.PERSISTENCE_CHECK_DURATION) {
-                        log.info(String.format(ERROR_EXCEED_DURATION, document, indexName));
+                if (Config.PERSISTENCE_CHECK_DURATION == 0) {
+                    continue;
+                }
```
Comment (damithc) [inline src/main/java/teammates/storage/search/SearchManager.java]: This still doesn't sound right. Probably this check should be done even before going inside the loop. i.e.
```
if (persistence check not needed)
    do a simple put
    return

do complicated put-with-retry
```

Comment (whipermr5) [inline src/main/java/teammates/storage/search/SearchManager.java]: Even if the persistence check is not needed, we still need the exponential backoff loop to handle transient backend errors ([reference](https://cloud.google.com/appengine/docs/standard/java/search/#Java_Putting_documents_in_an_index)).

Comment (damithc) [inline src/main/java/teammates/storage/search/SearchManager.java]: Oh? I didn't realize their backend requires retrying by default :-p

**Comment by junhaoyap** [comment]
Ping @damithc for next level approval

**Comment by ablyx** [comment]
@damithc 
> On a related note, I wonder if should put this in a queue and increase the retry count.

I've tried that on a separate pr, but it's causing the ui test to fail because the documents needs time to be added(since they are in a task queue) and thus they're not searchable. Should I proceed by perhaps trying to add a wait time in the test cases before searching for the documents?

**Comment by ablyx** [comment]
@damithc 
> On a related note, I wonder if should put this in a queue and increase the retry count.

I've tried that on a separate pr, but it's causing the ui test to fail because the documents needs time to be added(since they are in a task queue) and thus they're not searchable. Should I proceed by perhaps trying to add a wait time in the test cases before searching for the documents?

**Comment by damithc** [comment]
> I've tried that on a separate pr, but it's causing the ui test to fail because the documents needs time to be added(since they are in a task queue) and thus they're not searchable. Should I proceed by perhaps trying to add a wait time in the test cases before searching for the documents?

Can be done as a separate PR. Yes, might need waits, preferably, wait-till-success rather than fixed waits.

**Pull request merged**